### PR TITLE
Sas blob signing

### PR DESCRIPTION
**Pull request by kpfaulkner**
Added ability to create snapshots as per https://msdn.microsoft.com/en-us/library/azure/ee691971.aspx

### CODE REVIEW

File: storage/blob.go [hunk 1]
```diff
@@ -1263,16 +1272,31 @@ func (b BlobStorageClient) GetBlobSASURI(container, name string, expiry time.Tim
...
+}
+
+func blobSASStringToSign(signedVersion, canonicalizedResource, signedExpiry, signedPermissions string, signedIP string, protocols string) (string, error) {
 	var signedStart, signedIdentifier, rscc, rscd, rsce, rscl, rsct string
 
 	if signedVersion >= "2015-02-21" {
 		canonicalizedResource = "/blob" + canonicalizedResource
 	}
 
+	// https://msdn.microsoft.com/en-us/library/azure/dn140255.aspx#Anchor_8
```
Comment (kpfaulkner) [inline storage/blob.go]: done

File: storage/blob_test.go [hunk 2]
```diff
@@ -37,10 +37,10 @@ func (s *StorageBlobSuite) Test_pathForBlob(c *chk.C) {
 }
 
 func (s *StorageBlobSuite) Test_blobSASStringToSign(c *chk.C) {
```
Comment (kpfaulkner) [inline storage/blob_test.go]: Already existed in the TestGetBlobSASURIWithSignedIPAndProtocolValidAPIVersionPassed test, or did you want something else?

File: storage/blob_test.go [hunk 3]
```diff
@@ -71,6 +71,63 @@ func (s *StorageBlobSuite) TestGetBlobSASURI(c *chk.C) {
 	c.Assert(expectedParts.Query(), chk.DeepEquals, sasParts.Query())
 }
 
+// Newer version specified but cannot specify IP range since this isn't really a unit test
```
Comment (kpfaulkner) [inline storage/blob_test.go]: doh, you're right. added

File: storage/blob_test.go [hunk 4]
```diff
@@ -71,6 +71,63 @@ func (s *StorageBlobSuite) TestGetBlobSASURI(c *chk.C) {
...
+			"sp":  {"r"},
+			"se":  {"0001-01-01T00:00:00Z"},
+			"spr": {"https"},
+		}.Encode()}
+
+	u, err := cli.GetBlobSASURIWithSignedIPAndProtocol("container", "name", expiry, "r", "", []string{"https"})
+	c.Assert(err, chk.IsNil)
+	sasParts, err := url.Parse(u)
+	c.Assert(err, chk.IsNil)
+	c.Assert(sasParts.String(), chk.Equals, expectedParts.String())
```
Comment (kpfaulkner) [inline storage/blob_test.go]: My bad, though I did change to to sasParts.Query() and expectedParts.Query() as per pre-existing tests

File: storage/blob.go [hunk 5]
```diff
@@ -1210,11 +1208,13 @@ func pathForBlob(container, name string) string {
...
-// GetBlobSASURI creates an URL to the specified blob which contains the Shared
-// Access Signature with specified permissions and expiration time.
+// GetBlobSASURIWithSignedIPAndProtocol creates an URL to the specified blob which contains the Shared
+// Access Signature with specified permissions and expiration time. Also includes signedIPRange and allowed procotols.
+// If old API version is used but no signedIP is passed (ie empty string) then this should still work.
+// We only populate the signedIP when it non-empty.
 //
 // See https://msdn.microsoft.com/en-us/library/azure/ee395415.aspx
-func (b BlobStorageClient) GetBlobSASURI(container, name string, expiry time.Time, permissions string) (string, error) {
+func (b BlobStorageClient) GetBlobSASURIWithSignedIPAndProtocol(container, name string, expiry time.Time, permissions string, signedIPRange string, protocols []string) (string, error) {
```
Comment (kpfaulkner) [inline storage/blob.go]: done

File: storage/blob_test.go (hunk 2 above)
Comment (kpfaulkner) [inline storage/blob_test.go]: ahhh gotcha

**Comment by mcardosos** [comment]
LGTM, any more comments? @jhendrixMSFT , @vishrutshah  ?

**Comment by mcardosos** [comment]
Thanks for you contribution

**Pull request merged**